cdap-app-templates/cdap-program-report/cdap-program-report1_2.10/src/main/scala/co/cask/cdap/report/ReportAggregationFunction.scala

```diff
@@ -0,0 +1,240 @@
+    // Construct a status to time map from the list of status time tuples, by keeping the earliest time of a status
+    // if there exists multiple times for the same status
+    val statusTimeMap = bufferRow.getAs[Seq[Row]](STATUSES).groupBy(_.getAs[String](Constants.STATUS)).map(v =>
+      (v._1, v._2.map(_.getAs[Long](Constants.TIME)).min))
+
+    // get the status with maximum time as the status
+    val status = statusTimeMap.max(Ordering[Long].on[(_,Long)](_._2))._1
+    val start = statusTimeMap.get("STARTING")
+    val running = statusTimeMap.get("RUNNING")
+    // Get the earliest status with one of the ending statuses
+    val end = statusTimeMap.filterKeys(END_STATUSES.contains).values
+      .reduceOption(Math.min(_, _)) // avoid compilation error with Math.min(_, _) instead of Math.min
+    val startInfo = Option(bufferRow.getAs[Row](Constants.START_INFO))
+    val duration = end.flatMap(e => start.map(e - _))
+    val runtimeArgs = startInfo.map(_.getAs[Map[String, String]](Constants.RUNTIME_ARGUMENTS))
+    val startMethod = getStartMethod(runtimeArgs)
+    val r = Row(bufferRow.getAs[String](Constants.NAMESPACE),
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_NAME)).orNull,
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_VERSION)).orNull,
+      startInfo.map(_.getAs[String](Constants.ARTIFACT_SCOPE)).orNull,
+      bufferRow.getAs[String](Constants.APPLICATION_NAME),
+      bufferRow.getAs[String](Constants.APPLICATION_VERSION),
+      bufferRow.getAs[String](Constants.PROGRAM_TYPE), bufferRow.getAs[String](Constants.PROGRAM),
+      bufferRow.getAs[String](Constants.RUN), status,
+      start, running, end, duration, startInfo.map(_.getAs[String](Constants.USER)).orNull,
+      startMethod, runtimeArgs.orNull, 0, 0, 0)
+    LOG.trace("RecordBuilder = {}", buffer)
+    LOG.trace("Record = {}", r)
+    r
+  }
+
+  private def getStartMethod(runtimeArgs: Option[scala.collection.Map[String, String]]): String = {
+    if (runtimeArgs.isEmpty) return MANUAL
+    val scheduleInfoJson = runtimeArgs.get.get(SCHEDULE_INFO_KEY)
+    if (scheduleInfoJson.isEmpty) return MANUAL
+    val scheduleInfo: TriggeringScheduleInfo = GSON.fromJson(scheduleInfoJson.get, classOf[TriggeringScheduleInfo])
+    val triggers = scheduleInfo.getTriggerInfos
+    if (Option(triggers).isEmpty || triggers.isEmpty) return MANUAL
+    triggers.get(0).getType match {
+      case TriggerInfo.Type.TIME => SCHEDULED
+      case _ => TRIGGERED
+    }
+  }
+
+  private def rowToStartInfo(startInfoRow: Row): Row = {
+    val artifact: Row = startInfoRow.getAs[Row](Constants.ARTIFACT_ID)
+    Row(startInfoRow.getAs[String](Constants.USER),
+      startInfoRow.getAs[scala.collection.Map[String, String]](Constants.RUNTIME_ARGUMENTS),
+      artifact.getAs[String](Constants.ARTIFACT_NAME), artifact.getAs[String](Constants.ARTIFACT_VERSION),
+      artifact.getAs[String](Constants.ARTIFACT_SCOPE))
+  }
+}
+
```

Comment:
none of the methods have any docs, hard to read, can you add docs for these methods about what they do.